packages/nodes-base/nodes/Webhook/Webhook.node.ts

```diff
@@ -200,406 +86,206 @@ export class Webhook implements INodeType {
-								values: [
-									{
-										displayName: 'Name',
-										name: 'name',
-										type: 'string',
-										default: '',
-										description: 'Name of the header',
-									},
-									{
-										displayName: 'Value',
-										name: 'value',
-										type: 'string',
-										default: '',
-										description: 'Value of the header',
-									},
-								],
-							},
-						],
-					},
-					{
-						displayName: 'Property Name',
-						name: 'responsePropertyName',
-						type: 'string',
-						displayOptions: {
-							show: {
-								'/responseData': ['firstEntryJson'],
-								'/responseMode': ['lastNode'],
-							},
-						},
-						default: 'data',
-						description: 'Name of the property to return the data of instead of the whole JSON',
-					},
-				],
-			},
+			responseCodeProperty,
+			responseDataProperty,
+			responseBinaryPropertyNameProperty,
+			optionsProperty,
 		],
 	};
 
-	async webhook(this: IWebhookFunctions): Promise<IWebhookResponseData> {
-		const authentication = this.getNodeParameter('authentication') as string;
-		const options = this.getNodeParameter('options', {}) as IDataObject;
-		const req = this.getRequestObject();
-		const resp = this.getResponseObject();
-		const headers = this.getHeaderData();
-		const realm = 'Webhook';
+	async webhook(context: IWebhookFunctions): Promise<IWebhookResponseData> {
+		const options = context.getNodeParameter('options', {}) as {
+			binaryData: boolean;
+			ignoreBots: boolean;
+			rawBody: Buffer;
+			responseData?: string;
+		};
+		const req = context.getRequestObject();
+		const resp = context.getResponseObject();
 
-		const ignoreBots = options.ignoreBots as boolean;
-		if (ignoreBots && isbot((headers as IDataObject)['user-agent'] as string)) {
-			return authorizationError(resp, realm, 403);
+		try {
+			if (options.ignoreBots && isbot(req.headers['user-agent']))
+				throw new WebhookAuthorizationError(403);
+			await this.validateAuth(context);
+		} catch (error) {
+			if (error instanceof WebhookAuthorizationError) {
+				resp.writeHead(error.responseCode, { 'WWW-Authenticate': 'Basic realm="Webhook"' });
+				resp.end(error.message);
+				return { noWebhookResponse: true };
+			}
+			throw error;
 		}
 
+		const mimeType = req.headers['content-type'] ?? 'application/json';
+		if (mimeType.includes('multipart/form-data')) {
+			return this.handleFormData(context);
+		}
+
+		if (options.binaryData) {
+			return this.handleBinaryData(context);
+		}
+
+		const response: INodeExecutionData = {
+			json: {
+				headers: req.headers,
+				params: req.params,
+				query: req.query,
+				body: req.body,
+			},
+			binary: options.rawBody
+				? {
+						data: {
+							data: req.rawBody.toString(BINARY_ENCODING),
+							mimeType,
+						},
+				  }
+				: undefined,
+		};
+
+		return {
+			webhookResponse: options.responseData,
+			workflowData: [[response]],
+		};
+	}
+
+	private async validateAuth(context: IWebhookFunctions) {
```

Comment:
💥When running a Webhook with Basic Auth or Header Auth and you make a request with wrong credentials, an "_InternalServerError [ResponseError]: Cannot set headers after they are sent to the client_" will be thrown.

- On manual execution you'll receive "_Authorization data is wrong!_" but you can't continue with correct auth request.
- On production execution the n8n process witll crash

Reply:
I'm unable to reproduce this.

Reply:
never mind. found the issue, and it's fixed.